### PR TITLE
Initial pass at fragment interpolation

### PR DESCRIPTION
**Pull request by tgriesser**
Simple proof of concept for the ideas discussed in apollostack/eslint-plugin-graphql#19 - just to open up for discussion / feedback.

### CODE REVIEW

**Comment by stubailo** [comment]
See my comment here: https://github.com/apollostack/eslint-plugin-graphql/issues/19#issuecomment-254295602

I think using interpolation is not the right approach. Unfortunately, both Relay and Lokka use it, but I think it makes building great tools harder, and deviates from "just GraphQL".

And if we do build interpolation, I think it should preserve fragment names, so:

```js
const todoItem = gql`
  fragment todoItem on Todo {
    name
    completed
  }
`

gql`{
  todos {
    ...${todoItem}
  }
}`
```

Should result in:

```graphql
todos {
  ...todoItem
}

fragment todoItem on Todo {
  name
  completed
}
```

This will make it easier to analyze queries on the server-side, rather than losing the fragment name information. For example, it will be easier to find where that fragment came from in the client-side code.

**Comment by tgriesser** [comment]
> And if we do build interpolation, I think it should preserve fragment names, so:

Yep, this currently does do that, it just uses the [fragment name](https://github.com/apollostack/graphql-tag/pull/12/commits/a409bebfcd225a2d1d23885af01c3954b5b74ebe?diff=unified#diff-168726dbe96b3ce427e7fedce31bb0bcR85). The `...` is being added internally but there's no reason that couldn't change.

> I think using interpolation is not the right approach. Unfortunately, both Relay and Lokka use it, but I think it makes building great tools harder, and deviates from "just GraphQL".

I don't think it really deviates much from "just GraphQL", it just leverages the power of tagged templates to allow you to compose fragments rather than referencing them by a string name, which might not scale as well to larger codebases.

Also, If you need to change the fragment name, you currently need to update in every place it's used throughout the codebase / referenced in the query, rather than in just one place.

I am going to look a bit further into how difficult it'll be to get ESLint working with this.

**Comment by stubailo** [comment]
Oh, oops - I misread the code.

I think it would be fine to add interpolation features that preserve the fragment names, but I do think it makes it harder to generate stuff like persisted queries. If you have a build tool, let's say you see:

```js
const query = gql`
{
  todos {
    ${todoItem}
  }
}
`;
```

Now you need to understand the module system to understand what fields `todoItem` might have. If you don't use interpolation but use the fragment name directly, then it's really easy to find the matching fragment, wherever it may be in the code.

IMO having globally unique fragment names in the app makes like pretty easy. Let's say you have a fragment:

```js
const myFrag = gql`
  fragment todoItem on Todo { completed, text }
`;
```

Now if you rename that fragment, it's actually really easy to update all queries, you just search your codebase for `...todoItem`.

Again, with interpolation it's not that easy - you need to follow module imports/exports, etc.

Is there a benefit to using interpolation that I'm missing, over just using unique fragment names? From where I'm sitting, it seems like it actually makes it harder to work with the code.

**Comment by stubailo** [comment]
I guess the way I'm thinking about it is more similar to how the Facebook GraphQL tools work, as outlined here: https://github.com/graphql/graphql-js/pull/456

They just treat GraphQL queries and fragments as an app-global namespace, so that you can store and validate queries really easily without even looking at app code at all.

File: index.js [hunk 1]
```diff
@@ -49,9 +49,9 @@ function parseDocument(doc) {
 
   parsed = stripLoc(parsed);
 
-  cache[doc] = parsed;
+  cache[doc] = parsed.definitions;
```
Comment (stubailo) [inline index.js]: What's this change for?

File: index.js [hunk 2]
```diff
@@ -62,18 +62,48 @@ function gql(/* arguments */) {
...
   // run the loop only for the substitution count
   for (var i = 0; i < substitutions.length; i++) {
-      result += literals[i];
+    var substitution = substitutions[i]
+    var substitutionType = typeof substitution
+
+    result += literals[i];
+
+    if (substitutionType === 'string') {
+      console.warn('gql: Interpolation of arbitrary strings is deprecated')
```
Comment (stubailo) [inline index.js]: This could include a suggestion to use fragments instead.

File: index.js [hunk 3]
```diff
@@ -62,18 +62,48 @@ function gql(/* arguments */) {
...
       result += substitutions[i];
+    } else if (
+      substitutionType === 'object' && substitution.kind === 'Document'
+    ) {
+      var fragment = substitution.definitions[0]
```
Comment (stubailo) [inline index.js]: Documents can have multiple fragments, so we should either support that or error when there is more than one fragment.

File: index.js [hunk 4]
```diff
@@ -62,18 +62,48 @@ function gql(/* arguments */) {
...
+      result += '...' + fragment.name.value
+
+      for (var j = 0; j < substitution.definitions.length; j++) {
+        var frag = substitution.definitions[j]
+        if (fragments.indexOf(frag) === -1) {
+          fragments.push(frag)
+        }
+      }
+    } else {
+      throw new Error('gql: Invalid value interpolated in gql tagged template: ' + substitutionType)
```
Comment (stubailo) [inline index.js]: Instead of `invalid value`, we could say "Interpolation only supports strings and GraphQL document objects."

File: test.js [hunk 5]
```diff
@@ -8,8 +8,8 @@ var assert = require('chai').assert;
       assert.equal(gql`{ testQuery }`.kind, 'Document');
     });
 
-    it('returns the same object for the same query', () => {
```
Comment (stubailo) [inline test.js]: This was actually the intention - downstream tools use `===` to compare identical queries:

```js
gql`...` === gql`...`;
```

Why do we need to make this change?

**Comment by tgriesser** [comment]
Actually now that I've looked at the graphql-js issue you linked and looked a bit more into relay2 and how they're betting on fully static queries, I'm not sure this is the best idea. I think I'll look more into into approaches for that cli helper you mentioned.

**Pull request closed without merging**